### ISSUE TITLE
Publisher operators: where are we missing out?

### ISSUE THREAD

**Issue by klop** [pointfreeco/swift-composable-architecture#20]
I'd like to preface this issue by saying that I have not yet finished watching the episodes on TCA, and in fact I'm going to go back and watch them from the start very soon, so forgive me if I'm going wildly off course with this one.

Since I'm yet to come up with a well thought out description of the issue, and in fact may be prodding at something much different in scope to what my title suggests, I'll provide a concrete example: Cancellation.

There are a couple of ways through which effects can be cancelled in TCA:

1. `AnyCancellable` deinitizalization: stores subscribe to effects, effects return `AnyCancellable` when subscribed to. Those cancellables are held onto by the store, and if the store deinitializes so do the cancellables. `AnyCancellable` performs clean up on deinit.

2. Creating "cancellable" effects: effects can be made "cancellable" by calling the `cancellable(id: AnyHashable, cancelInFlight: Bool)` function on them. The `id` passed to `cancellable` can later be used to cancel the effect by calling the static `cancel(id: AnyHashable)` function, or by calling `cancellable(id: AnyHashable, cancelInFlight: Bool)` again with the same `id` and `cancelInFlight == true`, and then subscribing to that effect.

The second approach to effect cancellation, which is what I'd like to focus on, involves some additional state management behind the scenes which can be found in [Cancellation.swift](https://github.com/pointfreeco/swift-composable-architecture/blob/master/Sources/ComposableArchitecture/Effects/Cancellation.swift).

What stood out to me here is that we're replicating functionality that's (at least to some extent) achieved through various publisher flattening strategies. Let's suppose we have a button that when tapped, causes an API request to be made, the result of which is transformed into an action to update some state. There's also a caveat, which is that there should not be more than one API request happening at any one time.

This is simple enough to achieve in TCA: each button tap results in a new cancellable effect (created with `cancelInFlight == true`) being returned from a reducer, the result being that if a button is tapped while an API request is yet to complete, it will be cancelled and a new request will take its place. We can show this using a (bad) diagram where `o` denotes a value and `|` denotes completion:
```
Button taps:     ---o------------o----o-o---o---------->
API request #1:       ------o|
API request #2:                   ----
API request #3:                       --
API request #4:                         ----
API request #5:                             ------o|
```
Request 1 gets fired and completes, 2-4 all get fired but don't complete due to cancellation caused by button taps prior to completion, and 5 gets fired and completes.

The same sequence of API calls and the resulting output could instead be achieved using `switchToLatest` (`flatMapLatest` in RxSwift) n.b. pseudocode that has no obvious place in TCA:

```
didTapButton.switchToLatest {
    client.callSomeAPI()
}
```

Furthermore, let's suppose we want to slightly tweak the behaviour such that rather cancelling in flight API requests, we wait for them to complete before allowing new ones to be made. In the above example, we'd simply replace `switchToLatest` with `flatMapFirst` (that's RxSwift, it seems Combine doesn't have an equivalent yet). What about situations where we want to race inner publishers, merge them while limiting max concurrency, or throttle/debounce?

It isn't clear to me where these use cases fit into TCA.

**Comment by mbrandonw** [comment]
Hey @klop, thanks for the issue, and it's an important thing that hasn't been talked about on Point-Free yet, so you didn't miss anything. In an upcoming part of the tour series of episodes we are doing we will touch a bit on it, but it still won't quite answer your question, so I can try to do that here.

While reactive operators play an important role in TCA, they also play a slightly different role than if you are just using Combine/RxSwift unencumbered. The reason is that TCA demands that actions come in through the store, which leads to a reducer being called, which is just a plain function, and not apart of a reactive stream that you can `.switchToLatest` or any of those other fancy operators.

Sometimes operators can be used just as you would expect. For example, it is common to `.map` on an effect so that you can massage the effect's data into an action before returning from a reducer. Also if you needed to do any `.filter` logic on your effect that would work just fine. And even things like `zip` for running effects concurrently and `merge(...).take(1)` for racing effects work just fine.

And even long-living effects can use some of the more advanced operators, such as `switchToLatest`, `throttle`, etc. In fact, the [offline download demo](https://github.com/pointfreeco/swift-composable-architecture/blob/d2240d0e76c1a758dbadbf737ceefc888b2e807c/Examples/CaseStudies/SwiftUICaseStudies/04-HigherOrderReducers-ResuableOfflineDownloads/) has an [example](https://github.com/pointfreeco/swift-composable-architecture/blob/d2240d0e76c1a758dbadbf737ceefc888b2e807c/Examples/CaseStudies/SwiftUICaseStudies/04-HigherOrderReducers-ResuableOfflineDownloads/DownloadComponent.swift#L119-L121) of this. The `client.download` effect is a long living effect that sends a stream of download progress values (from a `URLSession`). These values come in rapidly, so we throttle them to 1 a second.

The main point is that some operators cannot be used as you might expect because you aren't operating on a publisher of states but rather you are returning a brand new publisher each time the reducer is invoked. And because of this, we create some helpers (like `.cancellable` and [`.debounce`](https://github.com/pointfreeco/swift-composable-architecture/blob/d2240d0e76c1a758dbadbf737ceefc888b2e807c/Examples/Search/Search/SearchView.swift#L73)) to help restore some of the ergonomics of the full-blown reactive style with the TCA style.

And to address your hypothetical change to the cancellation behavior:

> Furthermore, let's suppose we want to slightly tweak the behaviour such that rather cancelling in flight API requests, we wait for them to complete before allowing new ones to be made.

You would either need to cook up a new `Effect` helper effect to bake in that logic for you, or you would hold additional state in your feature so that you can perform this logic locally. It's possible, but different than what you might expect if you hope to use vanilla Combine.

There are some styles of architecture that prioritize the reactive elements over the reducer elements like we have in TCA. There is RxFeedback, ReactiveFeedback and CombineFeedback, and I enumerated some of the reasons we choose our style over those styles in an issue on another repo, [here](https://github.com/pointfreeco/episode-code-samples/issues/53).

So, tl;dr: while TCA uses reactive elements, it doesn't necessarily mean you can use _everything_ from the reactive artillery, and we have not spent much time discussing this aspect of TCA. 😀

**Comment by klop** [comment]
@mbrandonw thanks so much for the in depth response. Would it then be fair to say that Combine is used in TCA to control the output of effects, rather than the production of effects?

e.g. The [offline download demo](https://github.com/pointfreeco/swift-composable-architecture/blob/d2240d0e76c1a758dbadbf737ceefc888b2e807c/Examples/CaseStudies/SwiftUICaseStudies/04-HigherOrderReducers-ResuableOfflineDownloads/) example you provided demonstrates how the output (enum of progress, data etc.) of an effect can be controlled using `throttle`.

On the other hand, the classic [autocomplete search box example](https://github.com/ReactiveX/RxSwift/blob/master/Documentation/Why.md#transient-state) : 

```swift
searchTextField.rx.text
    .throttle(.milliseconds(300), scheduler: MainScheduler.instance)
    .flatMapLatest { query in
        API.getSearchResults(query)
    }
```

demonstrates how the production of effects can be controlled using `throttle` on a stream of text field values which are fed into `APIClient.getSearchResults`.

**Comment by mbrandonw** [comment]
> Would it then be fair to say that Combine is used in TCA to control the output of effects, but not the production of effects?

Yeah that is accurate way of putting it. Another way of saying it: Combine cannot be used to transform the actions sent to the store, e.g. if you have an action `.loginButtonTapped`, then Combine will not help you debounce/throttle/filter/etc the stream of values that occur as the user taps on the button.

> though I do wonder about ambiguity that comes along with having 2 methods called debounce on Effect that do different things, and how difficult it'd be to actually use throttle here should you require its behaviour.

Ambiguity is handled because our custom [`debounce`](https://github.com/pointfreeco/swift-composable-architecture/blob/d2240d0e76c1a758dbadbf737ceefc888b2e807c/Sources/ComposableArchitecture/Effects/Debouncing.swift#L26) operator takes an `id` parameter to identify the effect that you are debouncing.

Also, we plan on having `throttle` operator on `Effect` that works as you would expect, but we just haven't had time to work on that yet.

>  Not to say this has a place in TCA, just more of a thought as I continue to learn more about the architecture.

Yeah, I would say that last code snippet is akin to how RxFeedback/ReactiveFeedback handle effects, which is modeled as a `(AnyPublisher<State>) -> AnyPublisher<Action>` function.

**Comment by mbrandonw** [comment]
@klop do you have any other questions? If not, is it cool to close this issue?

**Comment by klop** [comment]
Feel free to close, thanks!
